### PR TITLE
ARROW-7405: [Java] ListVector isEmpty API is incorrect

### PR DESCRIPTION
**Pull request by tianchen92**
Related to [ARROW-7405](https://issues.apache.org/jira/browse/ARROW-7405).

Currently isEmpty API is always return false in BaseRepeatedValueVector, and its subclass ListVector did not overwrite this method.
This will lead to incorrect result, for example, a ListVector with data [1,2], null, [], [5,6] should get [false, false, true, false] with this API, but now it would return [false, false, false, false].

This change implements `isEmpty(int index)` for a `ListVector` that will return `true` if the index is an empty list or a null value, and `false` otherwise.

### CODE REVIEW

**Comment by liyafan82** [comment]
> Related to [ARROW-7405](https://issues.apache.org/jira/browse/ARROW-7405).
> 
> Currently isEmpty API is always return false in BaseRepeatedValueVector, and its subclass ListVector did not overwrite this method.
> This will lead to incorrect result, for example, a ListVector with data [1,2], null, [], [5,6] should get [false, false, true, false] with this API, but now it would return [false, false, false, false].

For [1,2], null, [], [5,6], it should return [false, true, true, false]? 
as in most scenarios, null should be considered empty?

**Comment by tianchen92** [comment]
> > Related to [ARROW-7405](https://issues.apache.org/jira/browse/ARROW-7405).
> > Currently isEmpty API is always return false in BaseRepeatedValueVector, and its subclass ListVector did not overwrite this method.
> > This will lead to incorrect result, for example, a ListVector with data [1,2], null, [], [5,6] should get [false, false, true, false] with this API, but now it would return [false, false, false, false].
> 
> For [1,2], null, [], [5,6], it should return [false, true, true, false]?
> as in most scenarios, null should be considered empty?

Thanks, I was also not sure about this and I have no strong insistence on this, let's see other's comments~

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java [hunk 1]
```diff
@@ -710,14 +710,29 @@ public Object getObject(int index) {
...
   public boolean isNull(int index) {
     return (isSet(index) == 0);
   }
 
+  /**
+   * Check if element at given index is empty list.
+   * @param index position of element
+   * @return true if element at given index is empty list, false otherwise
+   */
+  public boolean isEmpty(int index) {
```
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: @Override I'm surpised Error prone didn't can this.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java [hunk 2]
```diff
@@ -710,14 +710,29 @@ public Object getObject(int index) {
...
+    if (isSet(index) == 0) {
```
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: use isNull here.

**Comment by emkornfield** [comment]
>For [1,2], null, [], [5,6], it should return [false, true, true, false]?
as in most scenarios, null should be considered empty?

It might be worth a quick e-mail to the mailing list to see if anybody is used it at the expected behavior.  Given that is has been broken for this long, I'm not sure it makes too much of a difference. Either way we should be clarified in the parent class at least.

I'm not entirely sure the two should be conflated (we might want to throw an exception instead of returning a boolean) but I would need to think about this some more.

**Comment by tianchen92** [comment]
> > For [1,2], null, [], [5,6], it should return [false, true, true, false]?
> > as in most scenarios, null should be considered empty?
> 
> It might be worth a quick e-mail to the mailing list to see if anybody is used it at the expected behavior. Given that is has been broken for this long, I'm not sure it makes too much of a difference. Either way we should be clarified in the parent class at least.
> 
> I'm not entirely sure the two should be conflated (we might want to throw an exception instead of returning a boolean) but I would need to think about this some more.

ok, just opened a discuss :)

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java (hunk 1 above)
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: Thanks, fixed.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java (hunk 2 above)
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: thanks, fixed.

**Comment by BryanCutler** [comment]
I just responded to the thread. I don't feel strongly either way, so I agree treating null as empty sounds good to me. Does this need an update @tianchen92 ?

**Comment by tianchen92** [comment]
> I just responded to the thread. I don't feel strongly either way, so I agree treating null as empty sounds good to me. Does this need an update @tianchen92 ?

Sure, just updated, thanks!

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java [hunk 3]
```diff
@@ -950,6 +950,33 @@ public void testGetBufferSizeFor() {
...
+      assertFalse(vector.isNull(2));
+      assertTrue(vector.isEmpty(2));
+
+      assertFalse(vector.isEmpty(0));
+      assertFalse(vector.isEmpty(3));
+
+    }
+  }
+
+
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: nit: remove extra newlines

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java [hunk 4]
```diff
@@ -950,6 +950,33 @@ public void testGetBufferSizeFor() {
     }
   }
 
+  @Test
+  public void testIsEmpty() {
+    try (final ListVector vector = ListVector.empty("list", allocator)) {
+
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: and remove newline here

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java [hunk 5]
```diff
@@ -950,6 +950,33 @@ public void testGetBufferSizeFor() {
...
+      assertTrue(vector.isNull(1));
+      assertTrue(vector.isEmpty(1));
+
+      assertFalse(vector.isNull(2));
+      assertTrue(vector.isEmpty(2));
+
+      assertFalse(vector.isEmpty(0));
+      assertFalse(vector.isEmpty(3));
+
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: and remove newline here

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java [hunk 6]
```diff
@@ -950,6 +950,33 @@ public void testGetBufferSizeFor() {
...
+      writeIntValues(writer, new int[] {});
+      writeIntValues(writer, new int[] {5, 6});
+      writer.setValueCount(4);
+
+      assertTrue(vector.isNull(1));
+      assertTrue(vector.isEmpty(1));
+
+      assertFalse(vector.isNull(2));
+      assertTrue(vector.isEmpty(2));
+
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: and remove newline here

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java [hunk 7]
```diff
@@ -950,6 +950,33 @@ public void testGetBufferSizeFor() {
...
+      // set values [1,2], null, [], [5,6]
+      writeIntValues(writer, new int[] {1, 2});
+      writer.setPosition(2);
+      writeIntValues(writer, new int[] {});
+      writeIntValues(writer, new int[] {5, 6});
+      writer.setValueCount(4);
+
+      assertTrue(vector.isNull(1));
+      assertTrue(vector.isEmpty(1));
+
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: and remove newline here

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java [hunk 8]
```diff
@@ -950,6 +950,33 @@ public void testGetBufferSizeFor() {
...
+      assertFalse(vector.isNull(2));
+      assertTrue(vector.isEmpty(2));
+
+      assertFalse(vector.isEmpty(0));
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: please put the indicies your checking in order, so this should be first

**Review by BryanCutler — CHANGES_REQUESTED** [review]
LGTM, just some minor things to cleanup

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java [hunk 9]
```diff
@@ -716,14 +716,30 @@ public Object getObject(int index) {
...
    */
   @Override
   public boolean isNull(int index) {
     return (isSet(index) == 0);
   }
 
+  /**
+   * Check if element at given index is empty list.
+   * @param index position of element
+   * @return true if element at given index is empty list, false otherwise
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: maybe say "..is empty list or NULL, .."

**Comment by tianchen92** [comment]
> LGTM, just some minor things to cleanup

Thanks, done.

**Review by BryanCutler — APPROVED** [review]
LGTM

**Comment by BryanCutler** [comment]
merged to master, thanks @tianchen92 !

**Pull request closed without merging**